consumer/src/main/kotlin/streams/kafka/KafkaManualCommitEventConsumer.kt

```diff
@@ -1,17 +1,87 @@
 package streams.kafka
 
+import org.apache.kafka.clients.consumer.CommitFailedException
+import org.apache.kafka.clients.consumer.ConsumerRebalanceListener
 import org.apache.kafka.clients.consumer.OffsetAndMetadata
 import org.apache.kafka.common.TopicPartition
+import org.apache.kafka.common.errors.WakeupException
 import org.neo4j.logging.Log
+import streams.extensions.offsetAndMetadata
+import streams.extensions.topicPartition
 import streams.service.StreamsSinkEntity
-import streams.service.errors.ErrorService
+import java.time.Duration
 
-class KafkaManualCommitEventConsumer(private val config: KafkaSinkConfiguration,
-                                     private val log: Log): KafkaAutoCommitEventConsumer(config, log) {
+class KafkaManualCommitEventConsumer(config: KafkaSinkConfiguration,
+                                     private val log: Log,
+                                     topics: Set<String>): KafkaAutoCommitEventConsumer(config, log, topics) {
+
+    private val asyncCommit = config.streamsAsyncCommit
+
+    override fun stop() {
+        if (asyncCommit) {
+            doCommitSync()
+        }
+        super.stop()
+    }
+
+    private fun doCommitSync() {
+        try {
+            /*
+             * While everything is fine, we use commitAsync.
+             * It is faster, and if one commit fails, the next commit will serve as a retry.
+             * But if we are closing, there is no "next commit". We call commitSync(),
+             * because it will retry until it succeeds or suffers unrecoverable failure.
+             */
+            consumer.commitSync()
+        } catch (e: WakeupException) {
+            // we're shutting down, but finish the commit first and then
+            // rethrow the exception so that the main loop can exit
+            doCommitSync()
+            throw e
+        } catch (e: CommitFailedException) {
+            // the commit failed with an unrecoverable error. if there is any
+            // internal state which depended on the commit, you can clean it
+            // up here. otherwise it's reasonable to ignore the error and go on
+            log.debug("Commit failed", e)
```

Comment:
log.warn?   It may not be critical but user should know

Reply:
Done